lib/dependency-manager-adapters/base.js

```diff
@@ -0,0 +1,188 @@
+'use strict';
+
+const chalk = require('chalk');
+const debug = require('debug');
+const fs = require('fs-extra');
+const path = require('node:path');
+const semver = require('semver');
+const Backup = require('../utils/backup');
+const { LOCKFILE, PACKAGE_JSON } = require('../utils/package-managers');
+
+class BaseAdapter {
+  configKey = 'npm';
+  defaultInstallOptions = [];
+  lockfile = '';
+  name = '';
+  overridesKey = '';
+
+  backup = null;
+  debugFunction = null;
+
+  constructor(options) {
+    this.backup = new Backup({ cwd: options.cwd });
+    this.buildManagerOptions = options.buildManagerOptions;
+    this.cwd = options.cwd;
+    this.managerOptions = options.managerOptions;
+    this.run = options.run ?? require('../utils/run');
+  }
+
+  debug(...args) {
+    if (this.debugFunction === null) {
+      this.debugFunction = debug(`ember-try:dependency-manager-adapter:${this.name}`);
+    }
+
+    this.debugFunction(...args);
+  }
+
+  async setup(options = {}) {
+    this._checkForDifferentLockfiles(options.ui);
+
+    await this.backup.addFiles([PACKAGE_JSON, this.lockfile]);
+  }
+
+  async changeToDependencySet(dependencySet) {
+    await this.applyDependencySet(dependencySet);
+    await this._install(dependencySet);
+
+    const dependencies = {
+      ...dependencySet.dependencies,
+      ...dependencySet.devDependencies,
+    };
+
+    const currentDependencies = Object.keys(dependencies).map((name) => ({
+      name,
+      packageManager: this.name,
+      versionExpected: dependencies[name],
+      versionSeen: this._findCurrentVersionOf(name),
+    }));
+
+    this.debug('Switched to dependencies: \n', currentDependencies);
+
+    return currentDependencies;
+  }
+
+  async applyDependencySet(dependencySet) {
+    if (dependencySet === undefined) {
+      return;
+    }
+
+    this.debug('Changing to dependency set: %s', JSON.stringify(dependencySet));
+
+    const oldPackageJSON = JSON.parse(fs.readFileSync(this.backup.pathForFile(PACKAGE_JSON)));
+    const newPackageJSON = this._packageJSONForDependencySet(oldPackageJSON, dependencySet);
+
+    this.debug('Write package.json with: \n', JSON.stringify(newPackageJSON));
+
+    fs.writeFileSync(path.join(this.cwd, PACKAGE_JSON), JSON.stringify(newPackageJSON, null, 2));
+
+    // We restore the original lockfile here, so that we always create a minimal
+    // diff compared to the original locked dependency set:
+    await this.backup.restoreFile(this.lockfile);
```

Comment:
I think this is a significant change in behavior? Restoring the lockfile before the install is no different than leaving it in place and editing the package.json. 

It is debated often, but we've had everything 'float' in ember-try runs, both so that packages test against the newest versions of their in-range dependencies but also to help with situations where other deps need to float to enable the changed packages in the scenario.

Reply:
The pnpm adapter had this, so I thought it would make sense for all adapters. Will remove!

Reply:
ah weird, that must have snuck by